README.md
# ContactMe

Easy way to send an e-mail for your "contact me" web form.

## Features

* Remove undesireble characters (avoid script attacks)
* Validate client e-mail format
* Limit the number of e-mails from a client in a specific period
* E-mail encoded in base64
* Allow plain authentication with mail server
* Errors and warnings are logged in "/var/log/contactme.log" with fallback for standard output

## API

Expect a POST request containing the fields:

| Field   | Description                          |
| -----   | -----------                          |
| name    | Client name                          |
| email   | Client e-mail to return the contact  |
| subject | Subject of the client                |
| message | Message of the client                |

## Rate Limit

* Use the [token bucket](http://en.wikipedia.org/wiki/Token_bucket) strategy
* Rate limit per IP fixed in 5 e-mails per day (burst)
* Cleanup for entries older than a day (goroutine running every 5 minutes)

## HTTP status

| Status | Description                          |
| ------ | -----------                          |
| 200    | E-mail sent                          |
| 400    | Invalid client e-mail format         |
| 405    | Only POST requests are allowed       |
| 427    | Client already sent too many e-mails |
| 500    | Something went wrong in server-side  |

## Use it

This service has the following parameters to run:

| Parameter  | Environment Varible  | Description                                                      |
| ---------  | -------------------  | -----------                                                      |
| config     | CONTACTME_CONFIG     | Configuration file (other params have priority)                  |
| port       | CONTACTME_PORT       | Port to listen to (default: 80)                                  |
| mailserver | CONTACTME_MAILSERVER | E-mail server address with port (e.g. smtp.gmail.com:587)        |
| username   | CONTACTME_PASSWORD   | E-mail server authentication password (default: same of mailbox) |
| mailbox    | CONTACTME_MAILBOX    | E-mail address that will receive all the e-mails                 |

It is recommended before running the service to set the environment variables instead of using the
command line parameters for safety reasons (you don't want your password visible in the process
list), or use the configuration file.

Command line example (without using environment variables):

```
# contactme -s smtp.gmail.com:587 -p "crazypassword" -m my@email.com
```

To use the service with [Upstart](http://en.wikipedia.org/wiki/Upstart) you can generate the Debian
package with the script "gendeb.sh" (depends on [fpm](https://github.com/jordansissel/fpm)), install
it in your server, and fill the file "/etc/contactme/contactme.yaml" with your data.

```
# ./gendeb.sh 1.0 1
# dpkg -i contactme_1.0-1_amd64.deb
# service contactme start
```

## E-mail template

```
Client: <client's name>
-------------------------------------
<message>
-------------------------------------
E-mail sent via ContactMe.
http://github.com/rafaeljusto/contactme
```

## Client example

```html
<!doctype html>
<html>
  <head>
    <meta charset="utf-8"></meta>
    <style>
      input {
        display: block;
        margin-bottom: 20px;
      }

      input[type=submit] {
        margin: 20px auto 0px auto;
      }

      label {
        display: block;
      }
    </style>
  </head>
  <body>
    <form id="contactme">
      <fieldset>
        <legend>Contact Me</legend>

        <label for="name">Name</label>
        <input type="text" id="name" name="name" />

        <label for="email">E-mail</label>
        <input type="email" id="email" name="email" />

        <label for="subject">Subject</label>
        <input type="text" id="subject" name="subject" />

        <label for="message">Message</label>
        <textarea id="message" name="message" rows="10" cols="80"></textarea>
      </fieldset>

      <input type="submit" value="Send" />
    </form>

    <script src="//code.jquery.com/jquery-2.1.4.min.js"></script>
    <script type="text/javascript">
      $(function() {
        $("#contactme").submit(function(e) {
          e.preventDefault();

          var data = $(this).serialize();
          $.post("http://localhost", data)
            .done(function() {
              alert("E-mail sent!");
            })
            .fail(function() {
              alert("Error!");
            });
        });
      });
    </script>
  </body>
</html>
```


Changelog
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2015-05-08
### Added
- Using configuration file

## [1.0.0] - 2015-05-07
### Added
- Remove undesireble characters (avoid script attacks)
- Validate client e-mail format
- Limit the number of e-mails from a client in a specific period (rate limit)
- E-mail encoded in base64
- Allow plain authentication with mail server
- Errors and warnings are logged in "/var/log/contactme.log" with fallback for standard output

Godeps/_workspace/src/gopkg.in/yaml.v2/encode_test.go
package yaml_test

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/rafaeljusto/contactme/Godeps/_workspace/src/gopkg.in/yaml.v2"
	. "gopkg.in/check.v1"
	"net"
	"os"
)

var marshalIntTest = 123

var marshalTests = []struct {
	value interface{}
	data  string
}{
	{
		nil,
		"null\n",
	}, {
		&struct{}{},
		"{}\n",
	}, {
		map[string]string{"v": "hi"},
		"v: hi\n",
	}, {
		map[string]interface{}{"v": "hi"},
		"v: hi\n",
	}, {
		map[string]string{"v": "true"},
		"v: \"true\"\n",
	}, {
		map[string]string{"v": "false"},
		"v: \"false\"\n",
	}, {
		map[string]interface{}{"v": true},
		"v: true\n",
	}, {
		map[string]interface{}{"v": false},
		"v: false\n",
	}, {
		map[string]interface{}{"v": 10},
		"v: 10\n",
	}, {
		map[string]interface{}{"v": -10},
		"v: -10\n",
	}, {
		map[string]uint{"v": 42},
		"v: 42\n",
	}, {
		map[string]interface{}{"v": int64(4294967296)},
		"v: 4294967296\n",
	}, {
		map[string]int64{"v": int64(4294967296)},
		"v: 4294967296\n",
	}, {
		map[string]uint64{"v": 4294967296},
		"v: 4294967296\n",
	}, {
		map[string]interface{}{"v": "10"},
		"v: \"10\"\n",
	}, {
		map[string]interface{}{"v": 0.1},
		"v: 0.1\n",
	}, {
		map[string]interface{}{"v": float64(0.1)},
		"v: 0.1\n",
	}, {
		map[string]interface{}{"v": -0.1},
		"v: -0.1\n",
	}, {
		map[string]interface{}{"v": math.Inf(+1)},
		"v: .inf\n",
	}, {
		map[string]interface{}{"v": math.Inf(-1)},
		"v: -.inf\n",
	}, {
		map[string]interface{}{"v": math.NaN()},
		"v: .nan\n",
	}, {
		map[string]interface{}{"v": nil},
		"v: null\n",
	}, {
		map[string]interface{}{"v": ""},
		"v: \"\"\n",
	}, {
		map[string][]string{"v": []string{"A", "B"}},
		"v:\n- A\n- B\n",
	}, {
		map[string][]string{"v": []string{"A", "B\nC"}},
		"v:\n- A\n- |-\n  B\n  C\n",
	}, {
		map[string][]interface{}{"v": []interface{}{"A", 1, map[string][]int{"B": []int{2, 3}}}},
		"v:\n- A\n- 1\n- B:\n  - 2\n  - 3\n",
	}, {
		map[string]interface{}{"a": map[interface{}]interface{}{"b": "c"}},
		"a:\n  b: c\n",
	}, {
		map[string]interface{}{"a": "-"},
		"a: '-'\n",
	},

	// Simple values.
	{
		&marshalIntTest,
		"123\n",
	},

	// Structures
	{
		&struct{ Hello string }{"world"},
		"hello: world\n",
	}, {
		&struct {
			A struct {
				B string
			}
		}{struct{ B string }{"c"}},
		"a:\n  b: c\n",
	}, {
		&struct {
			A *struct {
				B string
			}
		}{&struct{ B string }{"c"}},
		"a:\n  b: c\n",
	}, {
		&struct {
			A *struct {
				B string
			}
		}{},
		"a: null\n",
	}, {
		&struct{ A int }{1},
		"a: 1\n",
	}, {
		&struct{ A []int }{[]int{1, 2}},
		"a:\n- 1\n- 2\n",
	}, {
		&struct {
			B int "a"
		}{1},
		"a: 1\n",
	}, {
		&struct{ A bool }{true},
		"a: true\n",
	},

	// Conditional flag
	{
		&struct {
			A int "a,omitempty"
			B int "b,omitempty"
		}{1, 0},
		"a: 1\n",
	}, {
		&struct {
			A int "a,omitempty"
			B int "b,omitempty"
		}{0, 0},
		"{}\n",
	}, {
		&struct {
			A *struct{ X, y int } "a,omitempty,flow"
		}{&struct{ X, y int }{1, 2}},
		"a: {x: 1}\n",
	}, {
		&struct {
			A *struct{ X, y int } "a,omitempty,flow"
		}{nil},
		"{}\n",
	}, {
		&struct {
			A *struct{ X, y int } "a,omitempty,flow"
		}{&struct{ X, y int }{}},
		"a: {x: 0}\n",
	}, {
		&struct {
			A struct{ X, y int } "a,omitempty,flow"
		}{struct{ X, y int }{1, 2}},
		"a: {x: 1}\n",
	}, {
		&struct {
			A struct{ X, y int } "a,omitempty,flow"
		}{struct{ X, y int }{0, 1}},
		"{}\n",
	},

	// Flow flag
	{
		&struct {
			A []int "a,flow"
		}{[]int{1, 2}},
		"a: [1, 2]\n",
	}, {
		&struct {
			A map[string]string "a,flow"
		}{map[string]string{"b": "c", "d": "e"}},
		"a: {b: c, d: e}\n",
	}, {
		&struct {
			A struct {
				B, D string
			} "a,flow"
		}{struct{ B, D string }{"c", "e"}},
		"a: {b: c, d: e}\n",
	},

	// Unexported field
	{
		&struct {
			u int
			A int
		}{0, 1},
		"a: 1\n",
	},

	// Ignored field
	{
		&struct {
			A int
			B int "-"
		}{1, 2},
		"a: 1\n",
	},

	// Struct inlining
	{
		&struct {
			A int
			C inlineB `yaml:",inline"`
		}{1, inlineB{2, inlineC{3}}},
		"a: 1\nb: 2\nc: 3\n",
	},

	// Map inlining
	{
		&struct {
			A int
			C map[string]int `yaml:",inline"`
		}{1, map[string]int{"b": 2, "c": 3}},
		"a: 1\nb: 2\nc: 3\n",
	},

	// Duration
	{
		map[string]time.Duration{"a": 3 * time.Second},
		"a: 3s\n",
	},

	// Issue #24: bug in map merging logic.
	{
		map[string]string{"a": "<foo>"},
		"a: <foo>\n",
	},

	// Issue #34: marshal unsupported base 60 floats quoted for compatibility
	// with old YAML 1.1 parsers.
	{
		map[string]string{"a": "1:1"},
		"a: \"1:1\"\n",
	},

	// Binary data.
	{
		map[string]string{"a": "\x00"},
		"a: \"\\0\"\n",
	}, {
		map[string]string{"a": "\x80\x81\x82"},
		"a: !!binary gIGC\n",
	}, {
		map[string]string{"a": strings.Repeat("\x90", 54)},
		"a: !!binary |\n  " + strings.Repeat("kJCQ", 17) + "kJ\n  CQ\n",
	},

	// Ordered maps.
	{
		&yaml.MapSlice{{"b", 2}, {"a", 1}, {"d", 4}, {"c", 3}, {"sub", yaml.MapSlice{{"e", 5}}}},
		"b: 2\na: 1\nd: 4\nc: 3\nsub:\n  e: 5\n",
	},

	// Encode unicode as utf-8 rather than in escaped form.
	{
		map[string]string{"a": "你好"},
		"a: 你好\n",
	},

	// Support encoding.TextMarshaler.
	{
		map[string]net.IP{"a": net.IPv4(1, 2, 3, 4)},
		"a: 1.2.3.4\n",
	},
	{
		map[string]time.Time{"a": time.Unix(1424801979, 0)},
		"a: 2015-02-24T18:19:39Z\n",
	},

	// Ensure strings containing ": " are quoted (reported as PR #43, but not reproducible).
	{
		map[string]string{"a": "b: c"},
		"a: 'b: c'\n",
	},
}

func (s *S) TestMarshal(c *C) {
	defer os.Setenv("TZ", os.Getenv("TZ"))
	os.Setenv("TZ", "UTC")
	for _, item := range marshalTests {
		data, err := yaml.Marshal(item.value)
		c.Assert(err, IsNil)
		c.Assert(string(data), Equals, item.data)
	}
}

var marshalErrorTests = []struct {
	value interface{}
	error string
	panic string
}{{
	value: &struct {
		B       int
		inlineB ",inline"
	}{1, inlineB{2, inlineC{3}}},
	panic: `Duplicated key 'b' in struct struct \{ B int; .*`,
}, {
	value: &struct {
		A int
		B map[string]int ",inline"
	}{1, map[string]int{"a": 2}},
	panic: `Can't have key "a" in inlined map; conflicts with struct field`,
}}

func (s *S) TestMarshalErrors(c *C) {
	for _, item := range marshalErrorTests {
		if item.panic != "" {
			c.Assert(func() { yaml.Marshal(item.value) }, PanicMatches, item.panic)
		} else {
			_, err := yaml.Marshal(item.value)
			c.Assert(err, ErrorMatches, item.error)
		}
	}
}

func (s *S) TestMarshalTypeCache(c *C) {
	var data []byte
	var err error
	func() {
		type T struct{ A int }
		data, err = yaml.Marshal(&T{})
		c.Assert(err, IsNil)
	}()
	func() {
		type T struct{ B int }
		data, err = yaml.Marshal(&T{})
		c.Assert(err, IsNil)
	}()
	c.Assert(string(data), Equals, "b: 0\n")
}

var marshalerTests = []struct {
	data  string
	value interface{}
}{
	{"_:\n  hi: there\n", map[interface{}]interface{}{"hi": "there"}},
	{"_:\n- 1\n- A\n", []interface{}{1, "A"}},
	{"_: 10\n", 10},
	{"_: null\n", nil},
	{"_: BAR!\n", "BAR!"},
}

type marshalerType struct {
	value interface{}
}

func (o marshalerType) MarshalText() ([]byte, error) {
	panic("MarshalText called on type with MarshalYAML")
}

func (o marshalerType) MarshalYAML() (interface{}, error) {
	return o.value, nil
}

type marshalerValue struct {
	Field marshalerType "_"
}

func (s *S) TestMarshaler(c *C) {
	for _, item := range marshalerTests {
		obj := &marshalerValue{}
		obj.Field.value = item.value
		data, err := yaml.Marshal(obj)
		c.Assert(err, IsNil)
		c.Assert(string(data), Equals, string(item.data))
	}
}

func (s *S) TestMarshalerWholeDocument(c *C) {
	obj := &marshalerType{}
	obj.value = map[string]string{"hello": "world!"}
	data, err := yaml.Marshal(obj)
	c.Assert(err, IsNil)
	c.Assert(string(data), Equals, "hello: world!\n")
}

type failingMarshaler struct{}

func (ft *failingMarshaler) MarshalYAML() (interface{}, error) {
	return nil, failingErr
}

func (s *S) TestMarshalerError(c *C) {
	_, err := yaml.Marshal(&failingMarshaler{})
	c.Assert(err, Equals, failingErr)
}

func (s *S) TestSortedOutput(c *C) {
	order := []interface{}{
		false,
		true,
		1,
		uint(1),
		1.0,
		1.1,
		1.2,
		2,
		uint(2),
		2.0,
		2.1,
		"",
		".1",
		".2",
		".a",
		"1",
		"2",
		"a!10",
		"a/2",
		"a/10",
		"a~10",
		"ab/1",
		"b/1",
		"b/01",
		"b/2",
		"b/02",
		"b/3",
		"b/03",
		"b1",
		"b01",
		"b3",
		"c2.10",
		"c10.2",
		"d1",
		"d12",
		"d12a",
	}
	m := make(map[interface{}]int)
	for _, k := range order {
		m[k] = 1
	}
	data, err := yaml.Marshal(m)
	c.Assert(err, IsNil)
	out := "\n" + string(data)
	last := 0
	for i, k := range order {
		repr := fmt.Sprint(k)
		if s, ok := k.(string); ok {
			if _, err = strconv.ParseFloat(repr, 32); s == "" || err == nil {
				repr = `"` + repr + `"`
			}
		}
		index := strings.Index(out, "\n"+repr+":")
		if index == -1 {
			c.Fatalf("%#v is not in the output: %#v", k, out)
		}
		if index < last {
			c.Fatalf("%#v was generated before %#v: %q", k, order[i-1], out)
		}
		last = index
	}
}


gendeb.sh
#!/bin/sh

usage() {
  echo "Usage: $1 <version> <release>"
}

pack_name="contactme"
version="$1"
release="$2"
vendor="Rafael Dantas Justo"
maintainer="Rafael Dantas Justo <adm@rafael.net.br>"
url="http://github.com/rafaeljusto/contactme"
license="MIT"
description="Microservice for sending e-mails via HTTP interface"

if [ -z "$version" ]; then
  echo "Version not defined!"
  usage $0
  exit 1
fi

if [ -z "$release" ]; then
  echo "Release not defined!"
  usage $0
  exit 1
fi

install_path=/usr/local/bin
config_path=/etc/contactme
tmp_dir=/tmp/contactme

workspace=`echo $GOPATH | cut -d: -f1`
workspace=$workspace/src/github.com/rafaeljusto/contactme

# recompiling everything
current_dir=`pwd`
cd $workspace
go build
cd $current_dir

if [ -f $pack_name*.deb ]; then
  # remove old deb
  rm $pack_name*.deb
fi

if [ -d $tmp_dir ]; then
  rm -rf $tmp_dir
fi

mkdir -p $tmp_dir$install_path $tmp_dir$config_path
mv $workspace/contactme $tmp_dir$install_path/
cp $workspace/contactme.yaml $tmp_dir$config_path/

fpm -s dir -t deb \
  --exclude=.git -n $pack_name -v "$version" --iteration "$release" --vendor "$vendor" \
  --maintainer "$maintainer" --url $url --license "$license" --description "$description" \
  --deb-upstart $workspace/contactme.upstart \
  --deb-user root --deb-group root \
  --prefix / -C $tmp_dir usr/local/bin etc/contactme



contactme.go
package main

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"net"
	"net/http"
	"net/mail"
	"net/smtp"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"text/template"
	"time"

	"github.com/rafaeljusto/contactme/Godeps/_workspace/src/github.com/codegangsta/cli"
	"github.com/rafaeljusto/contactme/Godeps/_workspace/src/gopkg.in/yaml.v2"
)

const (
	defaultPort               = 80
	defaultEmailSubjectPrefix = "[ContactMe] "
	defaultEmailTemplate      = `Client: {{.ClientName}}
-------------------------------------
{{.Message}}
-------------------------------------

E-mail sent via ContactMe.
http://github.com/rafaeljusto/contactme`
	defaultLog              = "/var/log/contactme.log"
	defaultRateLimitBurst   = 5.0
	defaultRateLimitRate    = 0.00035
	defaultRateLimitExpires = 25 * time.Hour
	defaultRateLimitCleanup = 5 * time.Minute

	// Possible exit codes on error
	errOpeningConfigFile    = 1
	errReadingConfigFile    = 2
	errSettingPort          = 3
	errParsingMailbox       = 4
	errMissingParameters    = 5
	errReadingEmailTemplate = 6
)

var (
	ratelimit     map[string]map[string]string
	ratelimitLock sync.RWMutex

	undesiredChars = regexp.MustCompile(`(['<>])|\\"|[^\x09\x0A\x0D\x20-\x7E\xA1-\xFF]`)

	config = struct {
		Port       int
		Mailserver struct {
			Address  string
			Username string
			Password string
		}
		Mailbox string
		Email   struct {
			SubjectPrefix string `yaml:"subject prefix"`
			Template      string
		}
		Log       string
		RateLimit struct {
			Burst   float64
			Rate    float64
			Expires time.Duration
			Cleanup time.Duration
		}
	}{
		Port: defaultPort,
		Email: struct {
			SubjectPrefix string `yaml:"subject prefix"`
			Template      string
		}{
			SubjectPrefix: defaultEmailSubjectPrefix,
			Template:      defaultEmailTemplate,
		},
		Log: "/var/log/contactme.log",
		RateLimit: struct {
			Burst   float64
			Rate    float64
			Expires time.Duration
			Cleanup time.Duration
		}{
			Burst:   defaultRateLimitBurst,
			Rate:    defaultRateLimitRate,
			Expires: defaultRateLimitExpires,
			Cleanup: defaultRateLimitCleanup,
		},
	}

	// Parsed template
	emailTemplate *template.Template
)

func init() {
	ratelimit = make(map[string]map[string]string)
}

func main() {
	app := cli.NewApp()
	app.Name = "contactme"
	app.Usage = "E-mail service with HTTP interface"
	app.Author = "Rafael Dantas Justo"
	app.Version = "0.1"

	app.Flags = []cli.Flag{
		cli.StringFlag{
			Name:   "config,c",
			EnvVar: "CONTACTME_CONFIG",
			Usage:  "Configuration file (other params have priority)",
		},
		cli.StringFlag{
			Name:   "port",
			EnvVar: "CONTACTME_PORT",
			Usage:  "Port to listen to (default: 80)",
		},
		cli.StringFlag{
			Name:   "mailserver,s",
			EnvVar: "CONTACTME_MAILSERVER",
			Usage:  "E-mail server address with port (e.g. smtp.gmail.com:587)",
		},
		cli.StringFlag{
			Name:   "username,u",
			EnvVar: "CONTACTME_USERNAME",
			Usage:  "E-mail server authentication username (default: same of mailbox)",
		},
		cli.StringFlag{
			Name:   "password,p",
			EnvVar: "CONTACTME_PASSWORD",
			Usage:  "E-mail server authentication password",
		},
		cli.StringFlag{
			Name:   "mailbox,m",
			EnvVar: "CONTACTME_MAILBOX",
			Usage:  "E-mail address that will receive all the e-mails",
		},
	}

	app.Action = func(c *cli.Context) {
		readCommandLineInputs(c)
		fillConfigurationDefaults()
		validateConfiguration()

		logFile := startLog()
		if logFile != nil {
			defer func() { logFile.Close() }()
		}

		go cleanup()

		http.HandleFunc("/", handle)
		log.Fatal(http.ListenAndServe(":"+strconv.Itoa(config.Port), nil))
	}

	app.Run(os.Args)
}

func readCommandLineInputs(c *cli.Context) {
	if configFile := c.String("config"); configFile != "" {
		data, err := ioutil.ReadFile(configFile)
		if err != nil {
			fmt.Printf("error opening configuration file. Details: %s\n", err)
			os.Exit(errOpeningConfigFile)
		}

		if err := yaml.Unmarshal(data, &config); err != nil {
			fmt.Printf("error reading configuration file. Details: %s\n", err)
			os.Exit(errReadingConfigFile)
		}
	}

	if port := c.String("port"); port != "" {
		var err error
		config.Port, err = strconv.Atoi(port)
		if err != nil {
			fmt.Printf("error setting port. Details: %s\n", err)
			os.Exit(errSettingPort)
		}
	}

	if mailserver := c.String("mailserver"); mailserver != "" {
		config.Mailserver.Address = mailserver
	}

	if username := c.String("username"); username != "" {
		config.Mailserver.Username = username
	}

	if password := c.String("password"); password != "" {
		config.Mailserver.Password = password
	}

	if mailbox := c.String("mailbox"); mailbox != "" {
		config.Mailbox = mailbox
	}
}

func fillConfigurationDefaults() {
	if config.Port == 0 {
		config.Port = defaultPort
	}

	config.Mailserver.Username = strings.TrimSpace(config.Mailserver.Username)
	if config.Mailserver.Username == "" {
		config.Mailserver.Username = config.Mailbox
	}

	config.Email.Template = strings.TrimSpace(config.Email.Template)
	if config.Email.Template == "" {
		config.Email.Template = defaultEmailTemplate
	}

	config.Log = strings.TrimSpace(config.Log)
	if config.Log == "" {
		config.Log = defaultLog
	}

	if config.RateLimit.Burst == 0 {
		config.RateLimit.Burst = defaultRateLimitBurst
	}

	if config.RateLimit.Rate == 0 {
		config.RateLimit.Rate = defaultRateLimitRate
	}

	if config.RateLimit.Expires.Seconds() == 0 {
		config.RateLimit.Expires = defaultRateLimitExpires
	}

	if config.RateLimit.Cleanup.Seconds() == 0 {
		config.RateLimit.Cleanup = defaultRateLimitCleanup
	}
}

func validateConfiguration() {
	config.Mailserver.Address = strings.TrimSpace(config.Mailserver.Address)
	config.Mailbox = strings.TrimSpace(config.Mailbox)

	if config.Mailserver.Address == "" || config.Mailbox == "" {
		fmt.Println("missing “mailserver” and/or “mailbox” arguments")
		os.Exit(errMissingParameters)
	}

	if _, err := mail.ParseAddress(config.Mailbox); err != nil {
		fmt.Printf("invalid mailbox “%s”\n", config.Mailbox)
		os.Exit(errParsingMailbox)
	}

	var err error
	emailTemplate, err = template.New("ContactMe E-mail").Parse(config.Email.Template)
	if err != nil {
		fmt.Printf("error reading e-mail template. Details: %s\n", err)
		os.Exit(errReadingEmailTemplate)
	}
}

func startLog() *os.File {
	logFile, err := os.OpenFile(config.Log, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
	if err != nil {
		fmt.Printf("error opening log file, "+
			"so we will use the standard output instead. Details: %s\n", err)

	} else {
		log.SetOutput(logFile)
	}

	return logFile
}

func handle(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "POST")

	if r.Method != "POST" {
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}

	ip, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		log.Println("invalid remote address. Details:", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	if granted, err := grant(ip); !granted {
		w.WriteHeader(427)
		return

	} else if err != nil {
		log.Println("error in rate limit. Details:", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	from, subject, body, err := readRequestInputs(r)
	if err != nil {
		log.Printf("invalid input from “%s”. Details: %s", ip, err)
		w.WriteHeader(http.StatusBadRequest)
		return
	}

	if err := sendEmail(from, subject, body); err != nil {
		log.Println("error sending e-mail. Details:", err)
		w.WriteHeader(http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}

func readRequestInputs(r *http.Request) (from, subject string, body bytes.Buffer, err error) {
	name := normalizeInput(r.FormValue("name"))
	from = normalizeInput(r.FormValue("email"))
	subject = normalizeInput(r.FormValue("subject"))
	message := normalizeInput(r.FormValue("message"))

	err = emailTemplate.Execute(&body, struct {
		ClientName string
		Message    string
	}{
		ClientName: name,
		Message:    message,
	})

	if err != nil {
		return
	}

	_, err = mail.ParseAddress(from)
	return
}

func sendEmail(from, subject string, body bytes.Buffer) error {
	header := map[string]string{
		"From":                      from,
		"To":                        config.Mailbox,
		"Subject":                   config.Email.SubjectPrefix + subject,
		"MIME-Version":              "1.0",
		"Content-Type":              `text/plain; charset="utf-8"`,
		"Content-Transfer-Encoding": "base64",
	}

	var message string
	for key, value := range header {
		message += fmt.Sprintf("%s: %s\r\n", key, value)
	}
	message += "\r\n" + base64.StdEncoding.EncodeToString(body.Bytes())

	var auth smtp.Auth
	if config.Mailserver.Username != "" && config.Mailserver.Password != "" {
		auth = smtp.PlainAuth("",
			config.Mailserver.Username,
			config.Mailserver.Password,
			config.Mailserver.Address[:strings.Index(config.Mailserver.Address, ":")],
		)
	}

	return smtp.SendMail(
		config.Mailserver.Address,
		auth,
		from,
		[]string{config.Mailbox},
		[]byte(message),
	)
}

func normalizeInput(input string) string {
	input = strings.TrimSpace(input)
	input = undesiredChars.ReplaceAllString(input, "")
	return input
}

func grant(ip string) (bool, error) {
	ratelimitLock.RLock()
	ratelimitItem := ratelimit[ip]
	ratelimitLock.RUnlock()

	now := time.Now().UTC()
	lastEvent := now

	if last, ok := ratelimitItem["last"]; ok {
		var err error
		lastEvent, err = time.Parse(time.RFC3339Nano, last)
		if err != nil {
			return false, err
		}
	}

	level := config.RateLimit.Burst

	if levelString, ok := ratelimitItem["level"]; ok {
		var err error
		level, err = strconv.ParseFloat(levelString, 64)
		if err != nil {
			return false, err
		}

		diff := now.Sub(lastEvent).Seconds()
		level = math.Min(config.RateLimit.Burst, level+diff*config.RateLimit.Rate)
	}

	answer := false

	if level >= 1.0 {
		level -= 1.0
		answer = true
	}

	if ratelimitItem == nil {
		ratelimitItem = make(map[string]string)
	}

	ratelimitItem["last"] = now.UTC().Format(time.RFC3339Nano)
	ratelimitItem["level"] = fmt.Sprintf("%f", level)

	ratelimitLock.Lock()
	ratelimit[ip] = ratelimitItem
	ratelimitLock.Unlock()

	return answer, nil
}

func cleanup() {
	for {
		newRatelimit := make(map[string]map[string]string)
		now := time.Now()

		ratelimitLock.Lock()
		for ip, ratelimitItem := range ratelimit {
			last, ok := ratelimitItem["last"]
			if !ok {
				newRatelimit[ip] = ratelimitItem
				continue
			}

			lastEvent, err := time.Parse(time.RFC3339Nano, last)
			if err != nil {
				newRatelimit[ip] = ratelimitItem
				continue
			}

			if now.Sub(lastEvent) <= config.RateLimit.Expires {
				newRatelimit[ip] = ratelimitItem
			}
		}
		ratelimit = newRatelimit
		ratelimitLock.Unlock()

		time.Sleep(config.RateLimit.Cleanup)
	}
}
